share/build-navigation/cache-request.xsl

```diff
@@ -0,0 +1,164 @@
+        </xsl:when>
+        <xsl:when test="$information = 'title'">
+          <xsl:call-template name="get-title">
+            <xsl:with-param name="productid" select="$productid"/>
+            <xsl:with-param name="setid" select="$setid"/>
+            <xsl:with-param name="doc_language" select="$doc_language"/>
+            <xsl:with-param name="dc" select="$dc"/>
+            <xsl:with-param name="rootid" select="$rootid"/>
+          </xsl:call-template>
+        </xsl:when>
+        <xsl:otherwise>
+          <xsl:message terminate="yes">Error: cache-request run without valid $information parameter.</xsl:message>
+        </xsl:otherwise>
+      </xsl:choose>
+    </xsl:variable>
+
+    <xsl:choose>
+      <xsl:when test="$result != ''">
+        <xsl:value-of select="$result"/>
+      </xsl:when>
+      <xsl:otherwise>
+        <xsl:message>Requested cached <xsl:value-of select="$information"/> for <xsl:value-of select="concat($dc,'/',$rootid)"/> does not exist.</xsl:message>
+      </xsl:otherwise>
+    </xsl:choose>
+  </xsl:template>
+
+
+  <xsl:template name="get-hash">
+    <xsl:param name="productid" select="''"/>
+    <xsl:param name="setid" select="''"/>
+    <xsl:param name="doc_language" select="''"/>
+    <xsl:param name="dc" select="''"/>
+    <xsl:param name="rootid" select="''"/>
+
+    <xsl:choose>
+      <xsl:when
+        test="$rootid != '' and
+              ($cache_content/document[@productid = $productid][@setid = $setid][@dc = $dc][@lang = $doc_language]/title[@rootid = $rootid]/@hash)[1]">
+        <xsl:value-of select="($cache_content/document[@productid = $productid][@setid = $setid][@dc = $dc][@lang = $doc_language]/title[@rootid = $rootid]/@hash)[1]"/>
+      </xsl:when>
+      <xsl:when test="($cache_content/document[@productid = $productid][@setid = $setid][@dc = $dc][@lang = $doc_language]/title[not(@rootid) or @rootid='']/@hash)[1]">
+        <xsl:value-of select="($cache_content/document[@productid = $productid][@setid = $setid][@dc = $dc][@lang = $doc_language]/title[not(@rootid) or @rootid='']/@hash)[1]"/>
+      </xsl:when>
+      <xsl:when test="($cache_content/document[@productid = $productid][@setid = $setid][@dc = $dc][@lang = $doc_language]/title/@hash)[1]">
+        <xsl:value-of select="($cache_content/document[@productid = $productid][@setid = $setid][@dc = $dc][@lang = $doc_language]/title/@hash)[1]"/>
+      </xsl:when>
+    </xsl:choose>
+  </xsl:template>
+
+  <xsl:template name="get-title">
+    <xsl:param name="productid" select="''"/>
+    <xsl:param name="setid" select="''"/>
+    <xsl:param name="doc_language" select="''"/>
+    <xsl:param name="dc" select="''"/>
+    <xsl:param name="rootid" select="''"/>
+
+    <xsl:variable name="title">
+      <xsl:choose>
+        <xsl:when
+          test="$rootid != '' and
+                ($cache_content/document[@productid = $productid][@setid = $setid][@dc = $dc][@lang = $doc_language]/title[@rootid = $rootid])[1]">
```

Comment:
These long variables seems a bit often repeated. Could you extract the part that is "constant" to all? Something like this:

```xml
<!-- maybe use a better name for the variable -->
<xsl:variable name="content" select="$cache_content/document[@productid = $productid][@setid = $setid][@dc = $dc][@lang = $doc_language]"/

<xsl:variable name="title">
   <xsl:choose>
      <xsl:when="$rootid != '' and ($content/title[@rootid = $rootid])[1]">
         <!-- ... -->
      </xsl:when>
   </xsl:choose>
</xsl:variable>
```

I think, that is better readable and should speed up things.